### ISSUE TITLE
Ability to send Aim data to remote storage

### ISSUE THREAD

**Issue by SGevorg** [aimhubio/aim#386]
Issue opened by almost everyone:
This is the very first feature towards the collaboration.
This issue captures the relatively complex feature to start with. 

We have been asked to build a number of different flavors of this. The discussion here and on the Slack will fully determine the first version of the feature to start with.

**Comment by SGevorg** [comment]
A link to a very informative thread on Reddit under Aim 2.2.0 launch about remote and potential bad behaviors to avoid:
https://www.reddit.com/r/MachineLearning/comments/mc8y1f/n_aim_220_is_out_hugging_face_integration_and/gs2li3h/?utm_source=reddit&utm_medium=web2x&context=3

**Comment by gorarakelyan** [comment]
**Overview**

Currently Aim collects and stores metadata on the training environment. MD(metadata) is collected via `Session class`. Each session creates a directory at the given path(repo + '/' + experiment) and stores MD within that directory(aka run). Additionally, Aim SDK gives flexibility to use multiple sessions in one training script to store multiple runs at once. 

> Check out more about Aim:
> - Overview -- https://github.com/aimhubio/aim#overview
> - Session class and specs -- https://github.com/aimhubio/aim#session
> - Basic example -- https://github.com/aimhubio/aim/blob/main/examples/sessions/multi_sessions.py

For example, when initializing a new Session via Aim SDK:
`my_sess = aim.Session(repo="/full/path/to/log/dir", experiment="my_experiment")`

`.aim` repo will have structure like this:
```
-> /full/path/to/log/dir
├── .aim  <--------------------- .aim directory where all the metadata is stored
│    ├── my_experiment  <------- the experiment name (experiments are used to group related runs together)
│    │   ├── aa-bbb-ccc-dd  <--- run associated with my_sess
│    │   │   ├── <all the MD associated with "aa-bbb-ccc-dd" run goes here -- created and tracked via my_sess>
│    │   ├── ...
│    │   └── aa-bbb-ccc-de  <--- other runs
│ 
│    ├── experiment_2  <- another experiment
│    │   └── aa-bbb-ccc-dd  <- a run related to the experiment_2
```

`my_sess` instance will inherit two main methods:
- [set_params](https://github.com/aimhubio/aim#set_params): Stores the params of the training run associated with the Session
- [track](https://github.com/aimhubio/aim#track): Tracks the training run metrics associated with the Session

Both of these methods store collected data at the `/full/path/to/log/dir + /.aim + /my_experiment + /aa-bbb-ccc-dd` path(where `aa-bbb-ccc-dd` is a random generated run uuid).

**Problem**

It would be great to provide an ability of sending and storing MD at a remote/centralized server.
Which would enable:
 - Saving aim logs in more secure and reliable place, rather than at the training environment
 - Sharing results with other team members
 - Ability of sending data from Colab and similar environments
 - Ability of sending data from local environments

**Solution**

Instead of storing MD on a local machine, serialized data could be sent to a remote server and stored there during the training process.

Session interface: `my_sess = aim.Session(repo="remote_server_url/log/dir/path", experiment...)`

Useful links:
- Session class source -- https://github.com/aimhubio/aim/blob/main/aim/sdk/session/session.py
- Track method -- https://github.com/aimhubio/aim/blob/main/aim/sdk/session/session.py
- Here artifacts are written to fs -- https://github.com/aimhubio/aim/blob/main/aim/sdk/session/session.py#L179
- Artifact, writers and serialization -- https://github.com/aimhubio/aim/tree/main/aim/artifacts

By modifying artifacts writing [here](https://github.com/aimhubio/aim/blob/main/aim/sdk/session/session.py#L179), we can send them to a remote server instead.

e.g.
```py
# Create and serialize artifact instance
artifact_inst = obj(*args, **kwargs, __aim_session_id=self.run_hash)

# ...

if remote_server:
   # send to remote server
else:
   writer = ArtifactWriter()
   writer.save(self.repo, artifact_inst)
```

_Note: Data loss should be prevented in case of network or electricity faults._

**Comment by fedeizzo** [comment]
Hi, I and my collegue @FrancescoBozzo have thought to a possible solution.

## Client side
User creates a session with a repo that is a url:
```python
aim_session = new Session(repo="http://myDomain.com")
```
Inside the costructor of Session class we can apply a regex or some other check (like isdir from os.path) in orderd to verify if the repo is a local path:
```python
def __init__(
    self, 
    repo: Optional[str] = None,
    ...
 ):
    ...
    if is_url(repo):
        self._remote_server = True
    else:
        self._remote_server = False
    ...
```
After session is created user calls track function during train loop:
```python
for e in epochs:
    ...
    aim_session.track(loss_value, name='loss', epoch=e, subset='train', dataset='train_1')
    ...
```
When track is called all methods are executed as usual until the ArtifactWriter invocation:
```python
if self._remote_server:
    writer = ArtifiactRemoteWriter()
    writer.save(self.repo, inst)
else:
    writer = ArtifactWriter()
    writer.save(self.repo, inst)
```
where ArtifactRemoteWriter is a new artifact that handles all http requests/reponses. Interfaces offered by this new Artifact will be the same of the ArtifactWriter

## Server side
A new folder called server will be created in the project structure:
```
aim
└── aim
    └── sdk
       ├── adapters
       ├── session
       └── server
```
Inside this folder a webserver will be created maybe with flask or fastAPI (tell us which you prefer). This webserver mantains a collections of Session inside a dictionary in which keys are string tokens and items are Session instances.
A possible API structure could be:
**Session creation**
```
PUT
http://myDomain.com:123/experiment_name/run_name
```
return value:
```json
{
  "experiment": "the_name_passed",
  "run": "the_name_passed",
  "id_token": "kl2190dsf908h123j4"
}
```
The token can be used to retrieve and update the session:
```python
a_session  = sessions_server_side[token] # instance of Session class
```
**Session tracking**
```
POST
http://myDomain.com:123/experiment_name/run_name
```
with a json body:
```json
{
  "id_token": "kl2190dsf908h123j4",
  ... all stuff passed to ArtifactRemoteWriters
}
```
Server handles request in this way:
```python
data = request.body
s = sessions_server_side.get(data.token, None)
if s is not None:
  s.track(data.other_stuff)
else:
  return an_error
```
and similar for other operations (close, flush, ...).

---

The dict object containg sessions will be periodically checked in order to verify if some client does not send data for a long period of time, if so server automatically flush and close the Session.
There will be a security mechanism which is ensured by an authentication layer.
Requests will be handled spawing threads.

Let me know if this solution is acceptable or if you prefer something else.

**Comment by gorarakelyan** [comment]
Hi @fedeizzo and @FrancescoBozzo,

Sounds awesome! Your solution perfectly fits.

I will share some thoughts regarding reliability, security and server API very shortly.

Thanks for the help guys 😊

**Comment by gorarakelyan** [comment]
@fedeizzo, @FrancescoBozzo,

Sharing some thoughts on above-mentioned topics:

## Reliability

### Client side fault tolerance: ability to respond to power and network failures, server errors and other faults

1. Using a file-based and process-safe queue to store all the session calls(`set_params`, `track`, `flush`, `close`). I've researched a bit and found some open source implementations. This solution looks good - https://dirq.readthedocs.io/en/latest/, but there might be more mature ones.
2. Executing a parallel thread which will continuously get items from the queue and send them to the server. It will ensure a solid level of tolerance.
For example, in case of:
    1. Network failures: the thread will wait until the connection is up again and continue sending items;
    2. Power failures: no data loss will occur, all the data is saved on the disk;
    3. Server errors and other faults: the thread will retry sending the packet at some interval(e.g. 2-5s.) until receiving `ok` status from the server.

_Note:_ A separate CLI command like `aim push` or `aim sync` should be provided to sync the queue with the server if the client machine was restarted. This can be added as an enhancement later.

### Server side fault tolerance: what happens if the server is stopped or restarted
The server should be able to recreate the session if the system was restarted. The idea of returning `token_id` is great and it also should include session information(args, kwargs) to give an ability of recreating session if it was destroyed or lost due to server faults. It can look like this:

```py
token_id = btoa(json.dump({
	'experiment': <experiment_name>, 
	'run': <run_hash>, 
	# **session_args, **kwargs
}))
```

If server receives `token_id` which can not be found in session instances, the session will be recreated, e.g.

```py
session_args = json.load(atob(token_id))
recreated_session = Session(session_args)
server_session_instances.append(recreated_session)
```

## Security

SSL or other protocols can be used for establishing encrypted links between server and client sides. It will ensure  authentication and transportation layers security. Can be implemented later as an enhancement.

## Server API
I was thinking about which transportation protocol would fit the best. Probably, websocket is a better fit, because the data is being continuously sent. But simple flask server is a good start. WS support can be added as a good addition later.

## Other notes
When initializing a remote session, an argument like`timeout` can be set to define server side session lifetime. By default it can be set to 1-2 hours. If server doesn't get any query within the given interval, the session object will be destroyed.

Does this make any sense to you? Happy to explore better solutions if you have any thoughts 😊

**Comment by fedeizzo** [comment]
Hi @gorarakelyan.

### Client side fault tolerance - dirq
Ok, no problem for this part. We think that the fs-queue will be implemented inside ArtifactRemoteWriter. In this way only when a remote repo is passed to the session the queue is created on disk.
We have some doubts on where queue folder will be saved on the file system. A temp dir inside /tmp or maye inside .aim folder?

Moreover we think that if a user specifies a url as repo string the automatic session init should be disabled in order to prevent .aim folder creation.

### Server side fault tolerance - token_id
The base64 encoding is a great idea. We found a problematic scenario:
(client and server up)

1. client sends session open messagge;
2. server opens the session and write down the folder in the repo;
3. client tracks some epochs;
4. server receives and saves them;
5. server crashes and loses opened sessions;
6. client continues to send data but fails;
7. server goes up again and receives track packets related to unknown sessions;

According to issue #403 there is no possibility to resume a session. When the server will try to open a new session decoding the token, it will find the experiment started during point 1. What should we do in this situation?

### Security
We have not experience in this field. Any clues?

### Server API
Sure, websocket are way better than http for this type of application. We are going to use [websockets](https://websockets.readthedocs.io/en/stable/intro.html)

**Comment by gorarakelyan** [comment]
Hi @fedeizzo ,

Thanks for the reply, I will answer you shortly.

**Comment by fedeizzo** [comment]
Hello,

at the moment we are busy with work and academic commitments, which is why our development is at a standstill.

When we will have more time, our idea is to open a PR with WIP tag (if you agree). This PR will be used to discuss some more difficult points. During these weeks we have realized that there is a greater amount of work than we thought. A few examples:

* the timestamp of a track operation is done inside the Session class. For this reason the data sent will be tagged when the remote repository saves them on disk and not when the user tracks them locally on his machine;
* currently the track function with epoch argument must be passed in ascending order (1, 2, 3 and so on). For this reason, the client must receive a notification of receipt from the server before sending any other data.

We think a PR is a good place to discuss similar things with the ability to highlight a few pieces of code. Let we know your opinion

**Comment by gorarakelyan** [comment]
Hi @fedeizzo ,

No worries at all. Taking into account that this is one of the most requested and needed features, it would be great if you could give even some broad estimations re timeline and the feature development cycles.

Yea, feel free to open a wip PR, defo a better place to discuss implementation details.

**Comment by gorarakelyan** [comment]
@fedeizzo, @FrancescoBozzo would love to invite you to join [Aim community slack](https://slack.aimstack.io) so we could communicate faster and more efficiently, while keeping main dev conversations and implementations details on GH.

**Comment by fru** [comment]
Hey, I wrote a little python project to create a http interface for aim. It tries to mimic the aim interface so it can be a drop in replacement: https://github.com/fru/aim-remote

**Comment by gorarakelyan** [comment]
@fru thanks a lot for the effort, that's amazing! 🚀 

The contributors have been working on the remote server for a couple of weeks now and it is planned to release an experimental PoC in the coming week. 
Will tag the contributors who are currently working on that part to go over the shared project and use/integrate it with the upcoming PoC if possible.

Also will make sure to share the next steps regarding taking it to a more stable and robust version. Will be super happy to help you contribute further on the project. ❤️ 

@mihran113 @alberttorosyan @devfox-se @mahnerak please take a look at the remote server implementation shared by @fru

**Comment by gorarakelyan** [comment]
@fedeizzo, @FrancescoBozzo the remote storage feature has been finally shipped with Aim v3.4! 🚀 

Please note that this is an experimental version, contributors are working hard to ship the stable version with the next release. 🤞 

See [the quick guide](https://aimstack.readthedocs.io/en/latest/guides/aim_remote_tracking.html) on how to track and send metadata to a remote server.

Feedbacks are welcome, don't hesitate to open an issue if you find any bugs! Thanks a lot!

**Comment by fedeizzo** [comment]
@gorarakelyan Glad to hear that, I will try the feature in the next few days

**Comment by roubkar** [comment]
Closing this issue due to inactivity.
